File: src/main/java/pt/joaomneto/titancompanion/adventure/impl/MRAdventure.java
package pt.joaomneto.titancompanion.adventure.impl;

import android.os.Bundle;
import android.view.Menu;

import java.io.BufferedWriter;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.List;

import pt.joaomneto.titancompanion.R;
import pt.joaomneto.titancompanion.adventure.SpellAdventure;
import pt.joaomneto.titancompanion.adventure.impl.fragments.mr.MRSkill;

public class MRAdventure extends SpellAdventure<MRSkill> {

    protected static final int FRAGMENT_SPELLS = 2;
    protected static final int FRAGMENT_EQUIPMENT = 3;
    protected static final int FRAGMENT_NOTES = 4;
    List<MRSkill> skills = new ArrayList<>();

    public MRAdventure() {
        super();
        fragmentConfiguration.clear();
        fragmentConfiguration.put(FRAGMENT_VITAL_STATS, new AdventureFragmentRunner(R.string.vitalStats,
                "pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureVitalStatsFragment"));
        fragmentConfiguration.put(FRAGMENT_COMBAT, new AdventureFragmentRunner(R.string.fights,
                "pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureCombatFragment"));
        fragmentConfiguration.put(FRAGMENT_SPELLS, new AdventureFragmentRunner(R.string.chosenSkills,
                "pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureSpellsFragment"));
        fragmentConfiguration.put(FRAGMENT_EQUIPMENT, new AdventureFragmentRunner(R.string.goldEquipment,
                "pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureEquipmentFragment"));
        fragmentConfiguration.put(FRAGMENT_NOTES, new AdventureFragmentRunner(R.string.notes,
                "pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureNotesFragment"));
    }

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        try {
            super.onCreate(savedInstanceState);

        } catch (Exception e) {
            e.printStackTrace();
        }

    }

    @Override
    public boolean onCreateOptionsMenu(Menu menu) {
        getMenuInflater().inflate(R.menu.adventure, menu);
        return true;
    }

    @Override
    public void storeAdventureSpecificValuesInFile(BufferedWriter bw) throws IOException {

        bw.write("skills=" + arrayToStringSpells(skills) + "\n");
        bw.write("gold=" + getGold() + "\n");
    }


    public List<MRSkill> getSpells() {
        return skills;
    }

    public void setSpells(List<MRSkill> skills) {
        this.skills = skills;
    }

    @Override
    protected void loadAdventureSpecificValuesFromFile() {
        setGold(Integer.valueOf(getSavedGame().getProperty("gold")));

        setSpells(stringToArraySpells(new String(getSavedGame().getProperty("skills").getBytes(java.nio.charset.Charset.forName("ISO-8859-1"))), MRSkill.class));

    }

    public List<MRSkill> getSpellList() {
        return Arrays.asList(MRSkill.values());
    }

    public boolean isSpellSingleUse() {
        return false;
    }

}

File: src/main/java/pt/joaomneto/titancompanion/adventure/impl/MRAdventure.kt
package pt.joaomneto.titancompanion.adventure.impl

import android.os.Bundle
import android.view.Menu
import pt.joaomneto.titancompanion.R
import pt.joaomneto.titancompanion.adventure.SpellAdventure
import pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureCombatFragment
import pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureEquipmentFragment
import pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureNotesFragment
import pt.joaomneto.titancompanion.adventure.impl.fragments.AdventureVitalStatsFragment
import pt.joaomneto.titancompanion.adventure.impl.fragments.mr.MRAdventureSpellsFragment
import pt.joaomneto.titancompanion.adventure.impl.fragments.mr.MRSkill
import pt.joaomneto.titancompanion.util.AdventureFragmentRunner
import java.io.BufferedWriter
import java.io.IOException
import java.util.ArrayList
import java.util.Arrays

class MRAdventure : SpellAdventure<MRSkill>(
    arrayOf(
        AdventureFragmentRunner(
            R.string.vitalStats,
            AdventureVitalStatsFragment::class
        ),
        AdventureFragmentRunner(
            R.string.chosenSkills,
            MRAdventureSpellsFragment::class
        ),
        AdventureFragmentRunner(
            R.string.fights,
            AdventureCombatFragment::class
        ),
        AdventureFragmentRunner(
            R.string.goldEquipment,
            AdventureEquipmentFragment::class
        ),
        AdventureFragmentRunner(
            R.string.notes,
            AdventureNotesFragment::class
        )
    )
) {

    internal var skills: MutableList<MRSkill> = ArrayList()

    val FRAGMENT_VITAL_STATS = 0
    val FRAGMENT_SPELLS = 1
    val FRAGMENT_COMBAT = 2
    val FRAGMENT_EQUIPMENT = 3
    val FRAGMENT_NOTES = 4

    override fun onCreate(savedInstanceState: Bundle?) {
        try {
            super.onCreate(savedInstanceState)
        } catch (e: Exception) {
            e.printStackTrace()
        }
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.adventure, menu)
        return true
    }

    @Throws(IOException::class)
    override fun storeAdventureSpecificValuesInFile(bw: BufferedWriter) {

        bw.write("skills=" + arrayToStringSpells(skills) + "\n")
        bw.write("gold=" + gold + "\n")
    }

    var spells: MutableList<MRSkill>
        get() = skills
        set(skills) {
            this.skills = skills
        }

    override fun loadAdventureSpecificValuesFromFile() {
        gold = Integer.valueOf(savedGame.getProperty("gold"))
        chosenSpells = stringToArraySpells(savedGame.getProperty("skills"), MRSkill::class.java)
    }

    override val spellList: MutableList<MRSkill>
        get() = Arrays.asList(*MRSkill.values())

    override val isSpellSingleUse: Boolean
        get() = false

    companion object
}
